### PR TITLE
chore: fix WTL warnings when building w GN

### PR DESCRIPTION
**Pull request by nornagon**
Chromium doesn't use this bit of WTL, so they haven't bothered to fix the warnings that are generated in these headers.

### CODE REVIEW

File: patches/common/chromium/wtl_warnings.patch [hunk 1]
```diff
@@ -0,0 +1,444 @@
...
+diff --git a/third_party/wtl/include/atldlgs.h b/third_party/wtl/include/atldlgs.h
+index 62db9178013c..4cf496eeea2d 100644
+--- a/third_party/wtl/include/atldlgs.h
++++ b/third_party/wtl/include/atldlgs.h
+@@ -473,7 +473,7 @@ public:
+ 		: CFileDialogImpl<T>(TRUE, lpszDefExt, lpszFileName, dwFlags, lpszFilter, hWndParent), 
+ 		  m_pNextFile(NULL)
+ 	{
+-		m_ofn.Flags |= OFN_ALLOWMULTISELECT;   // Force multiple selection mode
++		this->m_ofn.Flags |= OFN_ALLOWMULTISELECT;   // Force multiple selection mode
```
Comment (nornagon) [inline patches/common/chromium/wtl_warnings.patch]: This was the warning that triggered this (and all other similar changes):
```
../../third_party/wtl/include\atldlgs.h(476,3):  error: use of undeclared identifier 'm_ofn'; unqualified lookup into dependent bases of class template 'CMultiFileDialogImpl' is a Microsoft extension [-Werror,-Wmicrosoft-template]
                m_ofn.Flags |= OFN_ALLOWMULTISELECT;   // Force multiple selection mode
                ^
                this->
```

File: patches/common/chromium/wtl_warnings.patch [hunk 2]
```diff
@@ -0,0 +1,444 @@
...
++			ATLVERIFY(ResizeFilenameBuffer(this->m_ofn.nMaxFile + nExtraChars));
+ 	}
+ 
+ 	// Helper for _ATM_MIN_CRT
+@@ -1110,59 +1110,59 @@ public:
+ 	}
+ 
+ // Implementation - IFileDialogEvents interface
+-	virtual HRESULT STDMETHODCALLTYPE IFileDialogEvents::OnFileOk(IFileDialog* pfd)
++	virtual HRESULT STDMETHODCALLTYPE OnFileOk(IFileDialog* pfd)
```
Comment (nornagon) [inline patches/common/chromium/wtl_warnings.patch]: Fixes:
```
../../third_party/wtl/include\atldlgs.h(1113,55):  error: non-friend class member 'OnFileOk' cannot have a qualified name
        virtual HRESULT STDMETHODCALLTYPE IFileDialogEvents::OnFileOk(IFileDialog* pfd)
                                          ~~~~~~~~~~~~~~~~~~~^
../../third_party/wtl/include\atldlgs.h(1170,10):  error: 'WTL::CShellFileDialogImpl<WTL::CShellFileOpenDialog>::OnFileOk' hides overloaded virtual function [-Werror,-Woverloaded-virtual]
        HRESULT OnFileOk()
                ^
../../third_party/wtl/include\atldlgs.h(1211,55):  note: in instantiation of template class 'WTL::CShellFileDialogImpl<WTL::CShellFileOpenDialog>' requested here
class ATL_NO_VTABLE CShellFileOpenDialogImpl : public CShellFileDialogImpl< T >
                                                      ^
../../third_party/wtl/include\atldlgs.h(1238,37):  note: in instantiation of template class 'WTL::CShellFileOpenDialogImpl<WTL::CShellFileOpenDialog>' requested here
class CShellFileOpenDialog : public CShellFileOpenDialogImpl<CShellFileOpenDialog>
                                    ^
c:\program files (x86)\windows kits\10\include\10.0.15063.0\um\shobjidl_core.h(17698,43):  note: hidden overloaded virtual function 'IFileDialogEvents::OnFileOk' declared here: different number of parameters (1 vs 0)
        virtual HRESULT STDMETHODCALLTYPE OnFileOk(
                                          ^
```

File: patches/common/chromium/wtl_warnings.patch [hunk 3]
```diff
@@ -0,0 +1,444 @@
...
+ 	{
+ 		T* pT = static_cast<T*>(this);
+ 		ATLASSERT(pT->m_spFileDlg.IsEqualObject(pfd));
+-		pfd;   // avoid level 4 warning
++		(void)pfd;   // avoid level 4 warning
```
Comment (nornagon) [inline patches/common/chromium/wtl_warnings.patch]: Fixes:

```
../../third_party/wtl/include\atldlgs.h(1117,3):  error: expression result unused [-Werror,-Wunused-value]
                pfd;   // avoid level 4 warning
                ^~~
```

File: patches/common/chromium/wtl_warnings.patch [hunk 4]
```diff
@@ -0,0 +1,444 @@
...
+-			_Init(lpszFileName, dwOptions, lpszDefExt, arrFilterSpec, uFilterSpecCount);
++			this->_Init(lpszFileName, dwOptions, lpszDefExt, arrFilterSpec, uFilterSpecCount);
+ 	}
+ 
+ 	IFileOpenDialog* GetPtr()
+@@ -1244,6 +1244,7 @@ public:
+ 	                     const COMDLG_FILTERSPEC* arrFilterSpec = NULL, 
+ 	                     UINT uFilterSpecCount = 0U) : CShellFileOpenDialogImpl<CShellFileOpenDialog>(lpszFileName, dwOptions, lpszDefExt, arrFilterSpec, uFilterSpecCount)
+ 	{ }
++	virtual ~CShellFileOpenDialog() = default;
```
Comment (nornagon) [inline patches/common/chromium/wtl_warnings.patch]: Fixes a chromium-style issue (though that's weird, I thought chromium-style wasn't supposed to apply to things in `third_party/`...)

File: patches/common/chromium/wtl_warnings.patch [hunk 5]
```diff
@@ -0,0 +1,444 @@
...
++	virtual ~CShellFileSaveDialog() = default;
+ 
+ // Implementation (remove _Advise/_Unadvise code using template magic)
+ 	void _Advise(DWORD& /*dwCookie*/)
+@@ -2162,7 +2164,7 @@ public:
+ #ifndef _WIN32_WCE
+ 
+ // global helper
+-static HDC _AtlCreateDC(HGLOBAL hDevNames, HGLOBAL hDevMode)
++static inline HDC _AtlCreateDC(HGLOBAL hDevNames, HGLOBAL hDevMode)
```
Comment (nornagon) [inline patches/common/chromium/wtl_warnings.patch]: Fixes:
```
../../third_party/wtl/include\atldlgs.h(2167,12):  error: unused function '_AtlCreateDC' [-Werror,-Wunused-function]
static HDC _AtlCreateDC(HGLOBAL hDevNames, HGLOBAL hDevMode)
           ^
```

File: patches/common/chromium/wtl_warnings.patch [hunk 6]
```diff
@@ -0,0 +1,444 @@
...
+ {
+ 	if(hDevNames == NULL)
+ 		return NULL;
+@@ -2883,7 +2885,7 @@ public:
+ 		return hWnd;
+ 	}
+ 
+-	static const UINT GetFindReplaceMsg()
++	static UINT GetFindReplaceMsg()
```
Comment (nornagon) [inline patches/common/chromium/wtl_warnings.patch]: Fixes:
```
../../third_party/wtl/include\atldlgs.h(2888,9):  error: 'const' type qualifier on return type has no effect [-Werror,-Wignored-qualifiers]
        static const UINT GetFindReplaceMsg()
               ^~~~~~
```

File: patches/common/chromium/wtl_warnings.patch [hunk 7]
```diff
@@ -0,0 +1,444 @@
...
+diff --git a/third_party/wtl/include/atlgdi.h b/third_party/wtl/include/atlgdi.h
+index 510f1e7436f3..14c0b0eb0b41 100644
+--- a/third_party/wtl/include/atlgdi.h
++++ b/third_party/wtl/include/atlgdi.h
+@@ -3432,7 +3432,7 @@ public:
+ 	PIXELFORMATDESCRIPTOR m_pfd;
+ 
+ // Constructor/destructor
+-	CEnhMetaFileInfo(HENHMETAFILE hEMF) : m_pBits(NULL), m_pDesc(NULL), m_hEMF(hEMF)
++	CEnhMetaFileInfo(HENHMETAFILE hEMF) : m_hEMF(hEMF), m_pBits(NULL), m_pDesc(NULL)
```
Comment (nornagon) [inline patches/common/chromium/wtl_warnings.patch]: Fixes:
```
../../third_party/wtl/include/atlgdi.h(3435,55):  error: field 'm_pDesc' will be initialized after field 'm_hEMF' [-Werror,-Wreorder]
        CEnhMetaFileInfo(HENHMETAFILE hEMF) : m_pBits(NULL), m_pDesc(NULL), m_hEMF(hEMF)
                                                             ^
```

File: patches/common/chromium/wtl_warnings.patch [hunk 8]
```diff
@@ -0,0 +1,444 @@
...
+ 	{ }
+ 
+ 	~CEnhMetaFileInfo()
+@@ -3712,7 +3712,7 @@ inline HBITMAP AtlGetDibBitmap(LPBITMAPINFO pbmi)
+ 	void * pBits = NULL;
+ 
+ 	LPBYTE pDibBits = (LPBYTE)pbmi + sizeof(BITMAPINFOHEADER) + AtlGetDibColorTableSize(&pbmi->bmiHeader) * sizeof(RGBQUAD);
+-	if (hbm = CreateDIBSection(dc, pbmi, DIB_RGB_COLORS, &pBits, NULL, NULL)) 
++	if ((hbm = CreateDIBSection(dc, pbmi, DIB_RGB_COLORS, &pBits, NULL, NULL)))
```
Comment (nornagon) [inline patches/common/chromium/wtl_warnings.patch]: Fixes:
```
../../third_party/wtl/include/atlgdi.h(3715,10):  error: using the result of an assignment as a condition without parentheses [-Werror,-Wparentheses]
        if (hbm = CreateDIBSection(dc, pbmi, DIB_RGB_COLORS, &pBits, NULL, NULL))
            ~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../../third_party/wtl/include/atlgdi.h(3715,10):  note: place parentheses around the assignment to silence this warning
        if (hbm = CreateDIBSection(dc, pbmi, DIB_RGB_COLORS, &pBits, NULL, NULL))
                ^
            (                                                                   )
../../third_party/wtl/include/atlgdi.h(3715,10):  note: use '==' to turn this assignment into an equality comparison
        if (hbm = CreateDIBSection(dc, pbmi, DIB_RGB_COLORS, &pBits, NULL, NULL))
                ^
                ==
```

File: patches/common/chromium/wtl_warnings.patch [hunk 9]
```diff
@@ -0,0 +1,444 @@
...
+ 		memcpy(pBits, pDibBits, pbmi->bmiHeader.biSizeImage);
+ 
+ 	return hbm;
+@@ -3764,7 +3764,7 @@ inline HBITMAP AtlCopyBitmap(HBITMAP hbm , SIZE sizeDst, bool bAsBitmap = false)
+ 
+ inline HLOCAL AtlCreatePackedDib16(HBITMAP hbm, SIZE size)
+ {
+-	DIBSECTION ds = { 0 };
++	DIBSECTION ds = { { 0 } };
```
Comment (nornagon) [inline patches/common/chromium/wtl_warnings.patch]: Fixes:
```
../../third_party/wtl/include/atlgdi.h(3767,20):  error: suggest braces around initialization of subobject [-Werror,-Wmissing-braces]
        DIBSECTION ds = { 0 };
                          ^
                          {}
```

**Comment by deepak1556** [comment]
@nornagon any chance of upstreaming this patch ?

**Comment by nornagon** [comment]
See electron/electron#13463, we might be able to just drop it! 🎉

**Pull request merged**